routes/oauth.js

```diff
@@ -0,0 +1,65 @@
+"use strict";
+
+const AppConstants = require("../app-constants");
+
+const ClientOAuth2 = require("client-oauth2");
+const crypto = require("crypto");
+const express = require("express");
+const popsicle = require("popsicle");
+const router = express.Router();
+
+const gEmails = require("../subscribers");
+
+// This object exists instead of inlining the env vars to make it easy
+// to abstract fetching API endpoints from the OAuth server (instead
+// of specifying them in the environment) in the future.
+const FxAOAuthUtils = {
+  get authorizationUri() { return AppConstants.OAUTH_AUTHORIZATION_URI },
+  get tokenUri() { return AppConstants.OAUTH_TOKEN_URI },
+  get profileUri() { return AppConstants.OAUTH_PROFILE_URI },
+};
+
+var FxAOAuth = new ClientOAuth2({
+  clientId: AppConstants.OAUTH_CLIENT_ID,
+  clientSecret: AppConstants.OAUTH_CLIENT_SECRET,
+  accessTokenUri: FxAOAuthUtils.tokenUri,
+  authorizationUri: FxAOAuthUtils.authorizationUri,
+  redirectUri: AppConstants.SERVER_URL + "/oauth/redirect",
+  scopes: ["profile:email"],
+});
+
+router.get("/init", function(req, res) {
+  // Set a random state string in a cookie so that we can verify
+  // the user when they're redirected back to us after auth.
+  let state = crypto.randomBytes(40).toString("hex");
+  let uri = FxAOAuth.code.getUri({state});
+  req.session.state = state;
```

Comment:
We should make sure this session state cookie is set Secure and HTTP-only.

Also, does the server side of this state only exist in the node process memory? If so, that won't work when we scale this out to multiple processes/servers.

Reply:
Yeah, now that this refactor is done I'm going to start working on storage; this state as well as the email verification token should use a database.

Reply:
Filed #32